helix-core/src/selection.rs

```diff
@@ -728,6 +728,29 @@ pub fn split_on_matches(
     Selection::new(result, 0)
 }
 
+pub fn merge_consecutive_ranges(selection: &Selection) -> Selection {
+    let mut result = SmallVec::with_capacity(selection.len());
+    result.push(selection.ranges()[0]);
+
+    for sel in selection.ranges().get(1..).unwrap_or(&[]) {
+        let prev_sel: Range = match result.pop() {
+            None => {
+                break;
+            }
+            Some(s) => s,
+        };
+        if sel.from() == prev_sel.to() {
+            result.push(prev_sel.merge(*sel));
+        } else {
+            result.push(prev_sel);
+            result.push(*sel);
+        }
+    }
+
+    // TODO: figure out a new primary index
+    Selection::new(result, 0)
```

Comment:
You could use something similar to the code in normalize that also handles calculating the new primary index https://github.com/helix-editor/helix/blob/a4de86e7afc0da3d847d467a21ff4fb67a51d620/helix-core/src/selection.rs#L497-L522

Reply:
I'd make it modify a selection by value, since most code clones the selection first anyway. If would be almost the same code with the first conditional slightly changed

Reply:
Thanks for the suggestions. I added a calculation for the primary index, but I didn't understand the second comment very well, could you clarify what you mean?